### PR TITLE
Implement RFC5: State updates `Chain` type

### PR DESCRIPTION
**Pull request by yaahc**

### CODE REVIEW

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs]: This should be u128, for the same reasons as `Work`:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/work/difficulty.rs#L97

```suggestion
struct PartialCumulativeWork(u128);
```

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs]: Technically the number of transactions in a block is limited by the minimum transaction size, and maximum block size. But both those limits could change in future network upgrades. So I think `usize` is fine here.

```suggestion
    tx_by_hash: HashMap<transaction::Hash, (block::Height, usize)>,
```

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs]: The PartialEq, Eq, PartialOrd, and Ord implementations must be consistent.

Best practice is to write a single `cmp` function, then use it to implement all the other functions.
Here's an example of using `cmp` to implement `partial_cmp`:
https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html#how-can-i-implement-partialord

Here's an example of using `partial_cmp` to implement `eq`:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/work/difficulty.rs#L262
(But it should really be `cmp`.)

File: zebra-state/src/memory_state.rs
Comment (yaahc) [reply zebra-state/src/memory_state.rs]: I was thinking of adding a new type like `struct transparent::transaction::Index(usize);` as well

File: zebra-state/src/memory_state.rs
Comment (yaahc) [inline zebra-state/src/memory_state.rs]: It seems like `MemoryState` and `ChainSet` are ultimately two names for the same type, wondering if we should remove the `MemoryState` type or rename the `ChainSet` type and update the RFC.

Comment (teor2345) [reply zebra-state/src/memory_state.rs]: What types will we need to implement contextual validation?

I don't think `ChainSet` is a good name any more, because the type contains chains and queued blocks. What about `ChainState`?

(If we make that change, let's do a search and replace in the RFC.)

Here's one alternative, with a separate `QueuedBlocks` type:

```suggestion
/// A queue of blocks, awaiting the arrival of parent blocks.
pub struct QueuedBlocks {
    /// Blocks awaiting their parent blocks for contextual verification.
    blocks: BTreeMap<block::Hash, QueuedBlock>,
    /// Hashes from `queued_blocks`, indexed by parent hash.
    by_parent: BTreeMap<block::Hash, Vec<block::Hash>>,
    /// Hashes from `queued_blocks`, indexed by block height.
    by_height: BTreeMap<block::Height, Vec<block::Hash>>,
}

/// The state of the chains in memory, incuding queued blocks.
struct ChainState {
    /// Verified, non-finalized chains.
    chain_set: BTreeSet<Chain>,
    /// Blocks awaiting their parent blocks for contextual verification.
    contextual_queue: QueuedBlocks,
}
```

Here's a minimal change, which keeps a single type:

```suggestion
/// The state of the chains in memory, incuding queued blocks.
struct ChainState {
    /// Verified, non-finalized chains.
    chain_set: BTreeSet<Chain>,

    /// Blocks awaiting their parent blocks for contextual verification.
    queued_blocks: BTreeMap<block::Hash, QueuedBlock>,
    /// Hashes from `queued_blocks`, indexed by parent hash.
    queued_by_parent: BTreeMap<block::Hash, Vec<block::Hash>>,
    /// Hashes from `queued_blocks`, indexed by block height.
    queued_by_height: BTreeMap<block::Height, Vec<block::Hash>>,
}
```

File: book/src/dev/rfcs/0005-state-updates.md
Comment (teor2345) [inline book/src/dev/rfcs/0005-state-updates.md]: It's a BTreeMap, does it make sense to say "add to the end" ?

Also, it might be helpful to move block and work to the top, so it's very clear they are not part of `for each transaction`.

```suggestion
    - Add block to `self.blocks`
```

File: book/src/dev/rfcs/0005-state-updates.md
Comment (teor2345) [inline book/src/dev/rfcs/0005-state-updates.md]: I think this is slightly clearer:

Also, it might be helpful to move work to the top, so it's very clear it is not part of `for each transaction`.

```suggestion
    - Subtract work from `self.partial_cumulative_work`
```

File: book/src/dev/rfcs/0005-state-updates.md
Comment (teor2345) [inline book/src/dev/rfcs/0005-state-updates.md]:
```suggestion
    - Add block hash to `height_by_hash`
```

File: book/src/dev/rfcs/0005-state-updates.md
Comment (teor2345) [inline book/src/dev/rfcs/0005-state-updates.md]: It might be helpful to move work to the top, so it's very clear it is not part of `for each transaction`.

File: book/src/dev/rfcs/0005-state-updates.md
Comment (teor2345) [inline book/src/dev/rfcs/0005-state-updates.md]: `self.utxos` needs to be split into created and spent. Let's do a search and replace in the state RFC, so we catch all the remaining instances.

Also, the UTXOs, anchors, and nullifiers should be inside `for each transaction`.

File: book/src/dev/rfcs/0005-state-updates.md
Comment (teor2345) [inline book/src/dev/rfcs/0005-state-updates.md]: I just realised that `utxo_is_available` needs to be a per-chain API.

(Note that the Chain can be empty, if we are just starting or restarting.)

```suggestion
/// Is `utxo` available at the tip of `chain`?
utxo_is_available(utxo: UTXO, chain: Chain):
* is it in chain.spent? Return false
* is it in chain.created? Return true
* is it in the finalized state? Return true
* otherwise, Return false
```

File: zebra-chain/src/transaction.rs
Comment (teor2345) [inline zebra-chain/src/transaction.rs:99]: Is there a significant change here? Did I miss something?

File: zebra-chain/src/work/difficulty.rs
Comment (teor2345) [inline zebra-chain/src/work/difficulty.rs]: `usize` is 32-bit on some platforms. But like `Work`, `PartialCumulativeWork` needs to be `u128`.

The current Zcash mainnet does approximately 2^33 `Work` per second, and each block is approximately 75 seconds worth of `Work`. So each block has approximately 2^39 `Work`.

And we want to be able to scale to a much larger Zcash network in future. See this comment for more details:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/work/difficulty.rs#L107

```suggestion
pub struct PartialCumulativeWork(u128);
```

File: zebra-chain/src/work/difficulty.rs
Comment (teor2345) [inline zebra-chain/src/work/difficulty.rs]:
```suggestion
        let result = self.0
            .checked_add(rhs.0)
            .expect("Work values do not overflow")
```

File: zebra-chain/src/work/difficulty.rs
Comment (teor2345) [inline zebra-chain/src/work/difficulty.rs]: Subtractions can underflow - this is actually a very important check, because it makes sure we are only subtracting values we've previously added.

```suggestion
        let result = self.0
            .checked_sub(rhs.0)
            .expect("PartialCumulativeWork values do not underflow: all subtracted Work values must have been previously added to the PartialCumulativeWork")
```

File: zebra-state/src/lib.rs
Comment (teor2345) [inline zebra-state/src/lib.rs]: Can we make sure we re-enable these lints at some point?

Let's add them to the relevant tracking issue checklist.

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs]:
```suggestion
            .expect("only called while blocks is populated")
```

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs]:
```suggestion
            .expect("only called while blocks is populated");
```

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs]:
```suggestion
            .expect("only called while blocks is populated")
```

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs]:
```suggestion
            .expect("only called while blocks is populated");
```

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs]:
```suggestion
            .expect("only called while blocks is populated")
```

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs]: It might be helpful to call the argument `fork_tip`, to avoid confusion with the new block that's just arrived.

After the fork, we're going to append the new block to the forked chain, and it will become the new tip.
```suggestion
    pub fn fork(&self, fork_tip: block::Hash) -> Option<Self> {
        if !self.height_by_hash.contains_key(&fork_tip) {
            return None;
        }

        let mut forked = self.clone();

        while forked.non_finalized_tip_hash() != fork_tip {
```

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs:5]: Can we have a short module doc comment?

Let's also point to the state RFC doc.

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs:75]: I know doc comments aren't strictly required here, but short struct and function comments would be helpful.

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs:212]: 👍

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs:324]: 👍

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs]: The result of `PartialOrd` must be consistent with `Ord`.

In particular, we can't return `None` from `PartialOrd`, but `Some(_)` from `Ord` - implementing `Ord` guarantees that `PartialOrd` will never return `None`.

Also, using a single implementation avoids the risk of bugs, where we accidentally use `partial_cmp`, but expect a total order.

```suggestion
impl PartialOrd for Chain {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Chain {
    fn cmp(&self, other: &Self) -> Ordering {
        if self.partial_cumulative_work != other.partial_cumulative_work {
            self.partial_cumulative_work
                .cmp(&other.partial_cumulative_work)
        } else {
            let self_hash = self
                .blocks
                .values()
                .last()
                .expect("always at least 1 element")
                .hash();

            let other_hash = other
                .blocks
                .values()
                .last()
                .expect("always at least 1 element")
                .hash();

            // This comparison is a tie-breaker within the local node, so it does not need to
            // be consistent with the ordering on `ExpandedDifficulty` and `block::Hash`.
            match self_hash.0.cmp(&other_hash.0) {
                Ordering::Eq => unreachable!("Chain tip block hashes are always unique"),
                ordering @ _ => ordering,
            }
        }
    }
}
```

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs:168]: 👍 

Good idea, it is really helpful.

**Review by teor2345 — CHANGES_REQUESTED** [review]
Looks good, but needs a few tweaks and bug fixes.

I don't understand enough about transactions to review that part of the code, let's get @dconnolly or @hdevalence to review it as well.

In particular, I don't know if every post-Sapling block is guaranteed to have a Sprout and Sapling anchor. If they don't, that will affect this PR:
https://github.com/ZcashFoundation/zebra/pull/902#pullrequestreview-494520931

There are also a few outstanding design issues, which were part of my final review of the RFC, but didn't get included. I don't think they affect this PR, but please check:
https://github.com/ZcashFoundation/zebra/pull/1090

File: book/src/dev/rfcs/0005-state-updates.md
Comment (yaahc) [reply book/src/dev/rfcs/0005-state-updates.md]: I think moving the work to the top is a good idea but I'd prefer the block to be at the bottom because of ownership issues, if we insert it into the list immediately we have to make a clone which isn't a big deal but it seemed easy to avoid. The way I structured the implementation removes the manipulation of the block from the manipulation of the rest of the cummulative data members, so maybe it would be good to have both the "update cummulative work" and "for each transaction" bits indented and within an "update cummulative members" section. I'll reorganize things based on this and you let me know what you think.

File: book/src/dev/rfcs/0005-state-updates.md
Comment (yaahc) [reply book/src/dev/rfcs/0005-state-updates.md]: this is going to be an external API so we won't be using `Chain` as an input, probably `block::Hash` or the `Arc<Block>`? I'll check the async script RFC to make sure these are consistent.

File: zebra-chain/src/transaction.rs
Comment (yaahc) [reply zebra-chain/src/transaction.rs:99]: I talked to henry about doing this a while ago and just remembered to do it. The fact that `shielded_data` was above `joinsplit_data` despite being the newer member was confusing for me, I reordered this to reflect when they were added to the protocol.

File: zebra-state/src/lib.rs
Comment (yaahc) [reply zebra-state/src/lib.rs]: I'll add a checkbox to the original comment on this PR to remind me to remove these before we merge this.

File: zebra-state/src/memory_state.rs
Comment (yaahc) [reply zebra-state/src/memory_state.rs:5]: I'll draft this after I've written tests

File: zebra-state/src/memory_state.rs
Comment (yaahc) [reply zebra-state/src/memory_state.rs:75]: Sorry, I know I'm the worst about this, ty for the patient reminders :sweat_smile:

File: book/src/dev/rfcs/0005-state-updates.md
Comment (yaahc) [inline book/src/dev/rfcs/0005-state-updates.md:389]: @teor2345 I applied your change to split up the queued blocks here but I went ahead and switch it and the `SledState` to different names. lmk what you think

File: zebra-state/src/memory_state.rs
Comment (yaahc) [inline zebra-state/src/memory_state.rs:24]: @teor2345 I applied your change to split up the queued blocks here but I went ahead and switch it and the `SledState` to different names. lmk what you think

File: book/src/dev/rfcs/0005-state-updates.md
Comment (teor2345) [reply book/src/dev/rfcs/0005-state-updates.md]: Ok, so it will be passing the parent block hash instead

File: zebra-state/src/memory_state.rs
Comment (teor2345) [reply zebra-state/src/memory_state.rs:24]: Using the terms from the RFC is great!

File: book/src/dev/rfcs/0005-state-updates.md
Comment (teor2345) [inline book/src/dev/rfcs/0005-state-updates.md:297]: 👍

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs:148]: 👍

**Review by teor2345 — COMMENTED** [review]
Looks great!

I'm not sure if you'd like to merge now, or do some more work, and then merge.

I think it's ok to merge the partial PR, as long as someone else has reviewed the transaction parts.

**Comment by yaahc** [comment]
I want to get the two proptests that henry mentioned implemented before merging, just having a bit of trouble figuring out how to make arbitrary fake chains :sweat_smile:

**Comment by teor2345** [comment]
> I want to get the two proptests that henry mentioned implemented before merging, just having a bit of trouble figuring out how to make arbitrary fake chains 😅

Let me know if you need help with that - you might want to start at genesis, and then add various forks.

**Comment by teor2345** [comment]
I feel like this PR is now too big for me to review effectively.

Can we review `proptest-impl` separately, merge it, and then finish the reviews on this PR?

**Comment by yaahc** [comment]
@teor2345 https://discordapp.com/channels/676527656170160146/716086297210650634/758474653207101480

File: zebra-state/src/memory_state.rs
Comment (teor2345) [inline zebra-state/src/memory_state.rs]: I think you'll need to start chains with the genesis block, because parts of the state implementation expect that chains start at genesis.

**Comment by teor2345** [comment]
Just following up on the sprout and sapling roots - it doesn't look like we'll need to change the RFC design or implementation:
https://github.com/ZcashFoundation/zebra/pull/902/files#r493427948

File: book/src/dev/rfcs/0005-state-updates.md
Comment (teor2345) [inline book/src/dev/rfcs/0005-state-updates.md:789]: My original `utxo_is_available` was meant for contextual verification, not prospective verification. Since a chain can fork at any non-finalized block, I don't think we should check `spent_utxo` here.

Instead, we should check `spent_utxo` using the original `utxo_is_available` logic during contextual verification.

```suggestion
- (non-finalized) if any `Chains` contain `OutPoint` in their `created_utxos`, get the `transparent::Output` from `OutPoint`'s transaction
- (finalized) else if `OutPoint` is in `utxos_by_outpoint` return the associated `transparent::Output`.
```

**Review by teor2345 — APPROVED** [review]
I think we're looking pretty good to merge here.

We just need a tweak to the state design, and maybe we also want to get the separate tests PR working?

File: book/src/dev/rfcs/0005-state-updates.md
Comment (yaahc) [reply book/src/dev/rfcs/0005-state-updates.md:789]: applied this in the test PR to avoid needing a new test and approval

**Pull request merged**
